pkg/cafs/reader.go

```diff
@@ -201,115 +273,351 @@ func (r *chunkReader) WriteTo(writer io.Writer) (n int64, err error) {
+		leafPoolLogger.Debug("cafs freelist pool",
+			zap.Int("allocated buffers", r.leafPool.Buffers()),
+			zap.Int("free buffers", r.leafPool.FreeBuffers()),
+		)
+
+		for { // TODO(fred): nice - this could be chunked further and parallelized using rdr.ReadAt
+			select {
+			case <-doneC:
+				logger.Debug("interrupted readLeaf")
+				return nil, false, nil
+			default:
+			}
+
+			buffer := lb.Bytes()
+			read, e := rdr.Read(buffer[len(buffer):cap(buffer)])
+			if e != nil && e != io.EOF {
+				// relinquish trashed buffer
+				lb.Unpin()
+				r.leafPool.Release(lb)
+				return nil, false, e
 			}
-			lb.slice = lb.slice[:len(lb.slice)+m]
-			if e == io.EOF {
-				return lb, nil // e is EOF, so return nil explicitly
+
+			if read < 0 {
+				// safeguard against funny readers
+				panic(fmt.Errorf("reader returned a negative number of bytes"))
 			}
-			if e != nil {
-				r.leafPool.put(lb)
-				return lb, e
+
+			_ = lb.Slice(0, len(buffer)+read)
+			if e == io.EOF || read == 0 {
+				break
 			}
 		}
+		atomic.AddUint64(&r.fetched, 1)
+
+		if r.withVerifyHash {
+			logger.Debug("cafs reader ReadAt: hash verification")
+			// NOTE: we follow the checksumming scheme adopted by the writer
+			var (
+				i      int
+				isLast bool
+			)
+			if index+1 == len(r.keys) && uint32(len(lb.Bytes())) != r.leafSize {
+				i = index
+				isLast = true
+			} else {
+				i = index + 1
+			}
+			if err := r.verifyHash(k, lb.Bytes(), i, isLast); err != nil {
+				return nil, false, err
+			}
+		}
+		return lb, false, nil
 	}
+}
 
-	// Calculate first key and offset.
-	index, offset := calculateKeyAndOffset(off, r.leafSize)
-	if index >= int64(len(r.keys)) {
-		return 0, nil
-	}
-
-	var leafread int
-
-	addToCache := func(key Key, buffer *leafBuffer) {
+func addToCacheFunc(r *chunkReader) func(Key, LeafBuffer) {
+	return func(key Key, buffer LeafBuffer) {
 		if buffer == nil {
 			return
 		}
-		alreadyContained, _ := r.lru.ContainsOrAdd(key.StringWithPrefix(r.prefix), buffer)
+		alreadyContained, _ := r.lru.ContainsOrAdd(r.pather(key), buffer)
 		if alreadyContained {
-			r.leafPool.put(buffer)
+			// attention: we should only release buffers who where not already in the cache
+			//
+			// already in cache, relinquish freshly acquired buffer to pool
+			r.leafPool.Release(buffer)
 		}
 	}
+}
 
-	// seekAhead makes sure the next blob in CAFS is already in memory when the reader reaches that blob.
-	seekAhead := func() {
+// seekAhead prefetches blobs ahead
+func seekAheadFunc(r *chunkReader) func(int, int) bool {
+	return func(index, initiator int) bool {
 		nextIndex := index + 1
-		if nextIndex >= int64(len(r.keys)) {
-			return
+		if nextIndex >= len(r.keys) || nextIndex-initiator > r.maxFetchAhead {
+			// don't attempt to fetch farther ahead than permitted
+			return false
 		}
-		if r.lru.Contains(r.keys[nextIndex].StringWithPrefix(r.prefix)) {
-			return
+		if r.lru.Contains(r.pather(r.keys[nextIndex])) {
```

Comment:
again, suggest using `r.prefix` only, at least for now, until requirements necessitate configurable paths.

if deduplication (because the key here _must_ be the same as in the eviction callback) is the goal, i suggest naming a `func` at the package level and and using it in both places.

Reply:
the key is not used atm in the eviction callback, just the address of the buffer

Reply:
follow-up note: disagree - won't fix